memgpt/memory.py

```diff
@@ -16,96 +18,214 @@
     validate_date_format,
 )
 
-# from llama_index import Document
-# from llama_index.node_parser import SimpleNodeParser
 
+class MemoryModule(BaseModel):
+    """Base class for memory modules"""
 
-class CoreMemory(object):
-    """Held in-context inside the system message
+    description: Optional[str] = None
+    limit: int = 2000
+    value: Optional[Union[List[str], str]] = None
 
-    Core Memory: Refers to the system block, which provides essential, foundational context to the AI.
-    This includes the persona information, essential user details,
-    and any other baseline data you deem necessary for the AI's basic functioning.
-    """
+    def __setattr__(self, name, value):
+        """Run validation if self.value is updated"""
+        super().__setattr__(name, value)
+        if name == "value":
+            # run validation
+            self.__class__.validate(self.dict(exclude_unset=True))
 
-    def __init__(self, persona=None, human=None, persona_char_limit=None, human_char_limit=None, archival_memory_exists=True):
-        self.persona = persona
-        self.human = human
-        self.persona_char_limit = persona_char_limit
-        self.human_char_limit = human_char_limit
+    @validator("value", always=True)
+    def check_value_length(cls, v, values):
+        if v is not None:
+            # Fetching the limit from the values dictionary
+            limit = values.get("limit", 2000)  # Default to 2000 if limit is not yet set
 
-        # affects the error message the AI will see on overflow inserts
-        self.archival_memory_exists = archival_memory_exists
+            # Check if the value exceeds the limit
+            if isinstance(v, str):

```

Comment:
are these length values set in code downstream? or passed as instance values? if they are code this can just be a `field` prop (as a value, then I agree this is probably the best/most readable option)

Reply:
what do you mean by the length values? Both the value and potentially the limit (though we don't support this now) could be modified for an existing agent.

Reply:
Sorry, I was thinking ahead of my words with code - the "limit" values, not "length" values 🤦🏻 

basically, what I was trying to figure out by looking at the 2 classes is if one is a shell for the other and the limit is a default, or if the limit is the total that all the values must add up to. It's still not super clear (classic "naming things is the hardest part of programming" thing).  

Going through this PR review has been helpful - I think I get most of the relationships now, the values are intended to be the actual string content of a given section/partial of core memory, and it can be a single string or many strings that get concatenated into a single utterance within the context window. 

It would be solid to add this context into the objects (using class docstrings and `Field` description attributes) - it doesn't need to be in this PR. That's a great "first issue" to take the knowledge from these discussions and instrument the code base.